### PR TITLE
Redesign TypeSpec.Kind

### PR DESCRIPTION
**Pull request by Egorand**

### CODE REVIEW

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 1]
```diff
@@ -314,35 +307,51 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
...
+    internal val modifiers: MutableSet<KModifier> = mutableSetOf()
   ) {
-    CLASS(
+
+    val isEnum get() = this is Class && ENUM in modifiers
+
+    val isAnnotation get() = this is Class && ANNOTATION in modifiers
+
+    val isSimpleClass get() = this is Class && !isEnum && !isAnnotation
+
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Private?

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 2]
```diff
@@ -28,7 +33,6 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
   val anonymousTypeArguments = builder.anonymousTypeArguments
   val kdoc = builder.kdoc.build()
   val annotations = builder.annotations.toImmutableList()
-  val modifiers = builder.modifiers.toImmutableSet()
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: We still need to expose this (as Set). Delegate?

Comment (JakeWharton) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Actually we need a defensive copy of the kind or to make it immutable.

Comment (Egorand) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: It shouldn't be possible to modify the instance as all fields are internal, but I can look into improving encapsulation. `sealed` doesn't work with `data`, right?

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt (hunk 1 above)
Comment (Egorand) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Yes

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt (hunk 2 above)
Comment (JakeWharton) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: One builder creating two classes where the second adds a modifier will see that reflected on both.

Sealed works with data.

**Comment by Egorand** [comment]
@swankjesse yeah, I was expecting that, will rebase tomorrow
@JakeWharton `Kind` is now immutable (`modifiers` is a `Set`) and all fields are `internal`, making them `private` won't work since they're referenced in `Builder`

**Pull request merged**